### PR TITLE
Substitute QGL classes with QOpenGL

### PR DESCRIPTION
**Pull request by YakoYakoYokuYoku**
## PR Description

**What type of PR is this? (Check one of the boxes below)**

- [x] Bug fix (non-breaking change which fixes an issue)

**What does this pull request do?**

Given that the Qt OpenGL module was [deprecated in Qt 5.4.0](https://doc.qt.io/qt-5/qtopengl-index.html), removed in Qt 6 and most importantly was not functioning under a Wayland session (e.g. alpha output in KDE KWin or corrupted images in Sway), I've replaced its usage with the newer [OpenGL integration from Qt Gui](https://doc.qt.io/qt-5/qtgui-index.html#opengl-and-opengl-es-integration) and used both type definitions plus preprocessor conditionals for backwards compatibility with Qt 4.

**Show a few screenshots (if this is a visual change)**

**Have you tested your changes (if applicable)? If so, how?**

By building Natron and using the viewer, the dope sheet editor, the curve editor and the histogram.

**Futher details of this pull request**

Will post support for Wayland offscreen (`OSGLContext`) soon.

### CODE REVIEW

**Comment by devernay** [comment]
Did you test the Qt4 build as well?

**Comment by YakoYakoYokuYoku** [comment]
Yes, although it was only on a Linux box but I think it'll do fine on the other OSes.

File: Gui/QGLWidgetCompat.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Natron.  If not, see <http://www.gnu.org/licenses/gpl-2.0.html>
+ * ***** END LICENSE BLOCK ***** */
+
+#ifndef NATRON_GUI_QGLWIDGETCOMPAT_H
+#define NATRON_GUI_QGLWIDGETCOMPAT_H
+
+#include <QGLWidget>
```
Comment (devernay) [inline Gui/QGLWidgetCompat.h]: - add a comment at the top explaining what this is
- make sure GLIncludes.h is included before
- use the full include path `<QtOpenGL/QGLWidget>`
- OpenGL is deprecetad under macOS, silence it (see below)
- check that this is not included in Qt >= 5.4
```
#include "Global/Macros.h"
#include "Global/GLIncludes.h" //!<must be included before QGlWidget because of gl.h and glew.h
CLANG_DIAG_OFF(deprecated)
#include <QtOpenGL/QGLWidget>
CLANG_DIAG_ON(deprecated)
#if QT_VERSION >= QT_VERSION_CHECK(5, 4, 0)
#error "this file must not be included in Qt >= 5.4.0"
#endif
```

File: Gui/TextRenderer.cpp [hunk 2]
```diff
@@ -336,6 +344,21 @@ TextRenderer::renderText(float x,
                 assert( glIsTexture(texture) );
             }
             glCheckError();
+#if QT_VERSION >= QT_VERSION_CHECK(5, 4, 0)
+            glBegin(GL_TRIANGLES);
```
Comment (devernay) [inline Gui/TextRenderer.cpp]: What is the reason for not using `GL_QUADS` here? This looks like an unnecessary change.
Quads are perfectly valid: https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glBegin.xml

File: Gui/ViewerGL.cpp [hunk 3]
```diff
@@ -3617,7 +3631,11 @@ ViewerGL::swapOpenGLBuffers()
 {
     // always running in the main thread
     assert( qApp && qApp->thread() == QThread::currentThread() );
+#if QT_VERSION >= QT_VERSION_CHECK(5, 4, 0)
+    update();
```
Comment (devernay) [inline Gui/ViewerGL.cpp:3636]: Can you explain this change? I don't understand it.

File: Gui/QGLExtrasCompat.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Natron.  If not, see <http://www.gnu.org/licenses/gpl-2.0.html>
+ * ***** END LICENSE BLOCK ***** */
+
+#ifndef NATRON_GUI_QGLEXTRASCOMPAT_H
+#define NATRON_GUI_QGLEXTRASCOMPAT_H
+
+#include <QGLShaderProgram>
```
Comment (devernay) [inline Gui/QGLExtrasCompat.h]: Please include here the same changes as in QGLWidgetCompat.h

**Review by devernay — CHANGES_REQUESTED** [review]
asking for a few changes and explanations

File: Gui/ViewerGL.cpp (hunk 3 above)
Comment (devernay) [reply Gui/ViewerGL.cpp:3636]: ok, as I understand from the doc, no swapBuffers() in QOpenGLWidget, but shouldn't it call https://doc.qt.io/qt-5/qopenglcontext.html#swapBuffers instead?

File: Gui/TextRenderer.cpp (hunk 2 above)
Comment (YakoYakoYokuYoku) [reply Gui/TextRenderer.cpp]: This is more of a preference because using `GL_TRIANGLES` is more accurate than using `GL_QUADS` (see the [OpenGL wiki](https://www.khronos.org/opengl/wiki/Primitive#Quads) and this [StackOverflow thread](https://stackoverflow.com/questions/6644099/what-is-so-bad-about-gl-quads)). Could use `GL_QUADS` if that's wanted.

File: Gui/QGLWidgetCompat.h (hunk 1 above)
Comment (YakoYakoYokuYoku) [reply Gui/QGLWidgetCompat.h]: I went with checking that `GLIncludes.h` was before the `QGL*Compat.h` headers in the sources.

File: Gui/ViewerGL.cpp (hunk 3 above)
Comment (YakoYakoYokuYoku) [reply Gui/ViewerGL.cpp:3636]: In Qt5+ it's preferable to call `QOpenGLWidget::update`, which invokes `glSwapBuffers` (mentioned in the comment and in the docs), over to `QOpenGLContext::swapBuffers` in a Qt Widgets scenario because you have to pass a `QSurface` (e.g. `QWindow`) to the latter, which it's way more harder to setup and access in a widget application instead of a Qt Gui application.

File: Gui/TextRenderer.cpp (hunk 2 above)
Comment (devernay) [reply Gui/TextRenderer.cpp]: GL_QUADS is OK when the quad is planar, and even more when it's a rectangle, which is the case here.
So I don't see the point in rewriting it as GL_TRIANGLES, and having a different version for Qt5

**Comment by devernay** [comment]
I'll take some time to test on macOS (Qt4 and Qt5). Sorry if it causes some delay

File: Gui/TextRenderer.cpp (hunk 2 above)
Comment (YakoYakoYokuYoku) [reply Gui/TextRenderer.cpp]: With the deprecation of `QGLWidget::convertToGLFormat` (I think) the texture coordinates require to be flipped in the Y axis for accurate display of text through `Natron::TextRenderer`.

**Comment by YakoYakoYokuYoku** [comment]
> I'll take some time to test on macOS (Qt4 and Qt5). Sorry if it causes some delay

No problem with it and thanks for taking the time to test it.

File: Gui/TextRenderer.cpp (hunk 2 above)
Comment (devernay) [reply Gui/TextRenderer.cpp]: Yes maybe, according to https://stackoverflow.com/a/24907855

**Comment by YakoYakoYokuYoku** [comment]
@devernay how's testing going? As this is a blocker for some local work, although no problem if testing should take even more time.

**Comment by devernay** [comment]
good!

**Review by devernay — APPROVED** [review]
